File: frontend/src/lib/quantities.ts
/**
 * Unit conversions for bindable gauges (ADR 0039).
 *
 * SignalK is SI-canonical: oil pressure arrives in pascals, temperatures in
 * kelvin, fuel rate in cubic metres per second, engine hours in seconds, and
 * revolutions in hertz. None of those are readable on a helm display, and
 * `lib/units.ts` only knew feet and fahrenheit, so every gauge needs this.
 */

export type QuantityId =
  | 'pressure'
  | 'temperature'
  | 'volumetricFlow'
  | 'duration'
  | 'ratio'
  | 'volume'
  | 'speed'
  | 'length'
  | 'frequency'
  | 'raw'

export interface UnitOption {
  /** Stored in the widget config, so it must stay stable. */
  id: string
  label: string
  /** Converts from the SI base unit SignalK publishes. */
  fromSI: (value: number) => number
  decimals: number
}

export interface Quantity {
  id: QuantityId
  label: string
  /** SignalK's unit string for this quantity, used to infer it from meta. */
  siUnit: string
  units: UnitOption[]
}

const identity = (value: number) => value

export const QUANTITIES: Quantity[] = [
  {
    id: 'pressure',
    label: 'Pressure',
    siUnit: 'Pa',
    units: [
      { id: 'kPa', label: 'kPa', fromSI: (v) => v / 1000, decimals: 0 },
      { id: 'psi', label: 'psi', fromSI: (v) => v / 6894.757, decimals: 1 },
      { id: 'bar', label: 'bar', fromSI: (v) => v / 100000, decimals: 2 },
      { id: 'Pa', label: 'Pa', fromSI: identity, decimals: 0 },
    ],
  },
  {
    id: 'temperature',
    label: 'Temperature',
    siUnit: 'K',
    units: [
      { id: 'C', label: '°C', fromSI: (v) => v - 273.15, decimals: 1 },
      { id: 'F', label: '°F', fromSI: (v) => (v - 273.15) * 9 / 5 + 32, decimals: 1 },
      { id: 'K', label: 'K', fromSI: identity, decimals: 1 },
    ],
  },
  {
    id: 'volumetricFlow',
    label: 'Flow',
    siUnit: 'm3/s',
    units: [
      { id: 'Lph', label: 'L/h', fromSI: (v) => v * 1000 * 3600, decimals: 1 },
      { id: 'gph', label: 'gal/h', fromSI: (v) => v * 264.172 * 3600, decimals: 1 },
    ],
  },
  {
    id: 'duration',
    label: 'Duration',
    siUnit: 's',
    units: [
      { id: 'h', label: 'h', fromSI: (v) => v / 3600, decimals: 0 },
      { id: 'min', label: 'min', fromSI: (v) => v / 60, decimals: 0 },
      { id: 's', label: 's', fromSI: identity, decimals: 0 },
    ],
  },
  {
    id: 'ratio',
    label: 'Ratio',
    siUnit: 'ratio',
    units: [
      { id: 'percent', label: '%', fromSI: (v) => v * 100, decimals: 0 },
      { id: 'ratio', label: '', fromSI: identity, decimals: 2 },
    ],
  },
  {
    id: 'volume',
    label: 'Volume',
    siUnit: 'm3',
    units: [
      { id: 'L', label: 'L', fromSI: (v) => v * 1000, decimals: 0 },
      { id: 'gal', label: 'gal', fromSI: (v) => v * 264.172, decimals: 0 },
    ],
  },
  {
    id: 'speed',
    label: 'Speed',
    siUnit: 'm/s',
    units: [
      { id: 'kn', label: 'kts', fromSI: (v) => v * 1.943844, decimals: 1 },
      { id: 'kph', label: 'km/h', fromSI: (v) => v * 3.6, decimals: 1 },
      { id: 'mps', label: 'm/s', fromSI: identity, decimals: 1 },
    ],
  },
  {
    id: 'length',
    label: 'Length',
    siUnit: 'm',
    units: [
      { id: 'm', label: 'm', fromSI: identity, decimals: 1 },
      { id: 'ft', label: 'ft', fromSI: (v) => v * 3.28084, decimals: 1 },
    ],
  },
  {
    id: 'frequency',
    label: 'Frequency',
    siUnit: 'Hz',
    units: [
      // SignalK publishes revolutions in Hz; every tachometer on earth shows RPM.
      { id: 'rpm', label: 'RPM', fromSI: (v) => v * 60, decimals: 0 },
      { id: 'Hz', label: 'Hz', fromSI: identity, decimals: 1 },
    ],
  },
  {
    id: 'raw',
    label: 'Unitless',
    siUnit: '',
    units: [{ id: 'raw', label: '', fromSI: identity, decimals: 2 }],
  },
]

export function quantityById(id: QuantityId | string): Quantity {
  return QUANTITIES.find((quantity) => quantity.id === id) ?? QUANTITIES[QUANTITIES.length - 1]
}

/** Infers the quantity from SignalK's own meta.units, so the picker can preselect. */
export function quantityForSIUnit(siUnit: string | undefined): Quantity {
  const trimmed = (siUnit ?? '').trim()
  if (trimmed === '') return quantityById('raw')
  return QUANTITIES.find((quantity) => quantity.siUnit === trimmed) ?? quantityById('raw')
}

export function unitOption(quantityId: QuantityId | string, unitId: string | undefined): UnitOption {
  const quantity = quantityById(quantityId)
  return quantity.units.find((unit) => unit.id === unitId) ?? quantity.units[0]
}

/**
 * Converts and formats a raw SI value for display. Returns null when there is
 * nothing to show, so callers render the structural dash rather than a zero —
 * a gauge reading 0 when it means "no data" is the dangerous failure.
 */
export function formatQuantity(
  value: number | null | undefined,
  quantityId: QuantityId | string,
  unitId: string | undefined,
  decimals?: number,
): string | null {
  if (value === null || value === undefined || !Number.isFinite(value)) return null

  const unit = unitOption(quantityId, unitId)
  const converted = unit.fromSI(value)
  const places = decimals ?? unit.decimals

  return converted.toFixed(Math.max(0, Math.min(6, places)))
}

export function convertFromSI(value: number, quantityId: QuantityId | string, unitId: string | undefined): number {
  return unitOption(quantityId, unitId).fromSI(value)
}

File: backend/quantities.go
package main

import "fmt"

// Unit conversion, mirroring frontend/src/lib/quantities.ts in the toSI
// direction.
//
// The backend needed none of this until gauge zones became the alarm source
// (ADR 0050). Zones are authored in display units — oil pressure in psi,
// temperature in °C — because that is what the operator sets them against on
// the gauge scale. alarmReader reads SI from the SignalK snapshot: pascals,
// kelvin. Comparing one against the other produces an alarm that never fires
// and says nothing about why, so the threshold has to be converted once, here.
//
// Both directions are defined per unit and round-trip-tested against each
// other, and the table is asserted against the same values as the frontend's
// quantities.test.ts, so the two cannot drift apart silently.

type siUnitOption struct {
	ID string
	// FromSI matches the frontend's fromSI exactly; ToSI is its inverse.
	FromSI func(float64) float64
	ToSI   func(float64) float64
}

type siQuantity struct {
	ID    string
	Units []siUnitOption
}

func siIdentity(v float64) float64 { return v }

// linear builds the common case: a unit that is the SI value scaled by a factor.
func linear(id string, perSI float64) siUnitOption {
	return siUnitOption{
		ID:     id,
		FromSI: func(v float64) float64 { return v * perSI },
		ToSI:   func(v float64) float64 { return v / perSI },
	}
}

func identityUnit(id string) siUnitOption {
	return siUnitOption{ID: id, FromSI: siIdentity, ToSI: siIdentity}
}

var siQuantities = []siQuantity{
	{ID: "pressure", Units: []siUnitOption{
		linear("kPa", 1.0/1000),
		linear("psi", 1.0/6894.757),
		linear("bar", 1.0/100000),
		identityUnit("Pa"),
	}},
	{ID: "temperature", Units: []siUnitOption{
		{ID: "C",
			FromSI: func(v float64) float64 { return v - 273.15 },
			ToSI:   func(v float64) float64 { return v + 273.15 }},
		{ID: "F",
			FromSI: func(v float64) float64 { return (v-273.15)*9/5 + 32 },
			ToSI:   func(v float64) float64 { return (v-32)*5/9 + 273.15 }},
		identityUnit("K"),
	}},
	{ID: "volumetricFlow", Units: []siUnitOption{
		linear("Lph", 1000*3600),
		linear("gph", 264.172*3600),
	}},
	{ID: "duration", Units: []siUnitOption{
		linear("h", 1.0/3600),
		linear("min", 1.0/60),
		identityUnit("s"),
	}},
	{ID: "ratio", Units: []siUnitOption{
		linear("percent", 100),
		identityUnit("ratio"),
	}},
	{ID: "volume", Units: []siUnitOption{
		linear("L", 1000),
		linear("gal", 264.172),
	}},
	{ID: "speed", Units: []siUnitOption{
		linear("kn", 1.943844),
		linear("kph", 3.6),
		identityUnit("mps"),
	}},
	{ID: "length", Units: []siUnitOption{
		identityUnit("m"),
		linear("ft", 3.28084),
	}},
	// Distance per unit volume, which is how SignalK states fuel economy.
	{ID: "fuelEconomy", Units: []siUnitOption{
		linear("nmpl", 1.0/1852000),
		linear("nmpg", 3.785411784/1852000),
		identityUnit("m/m3"),
	}},
	{ID: "power", Units: []siUnitOption{
		linear("kW", 1.0/1000),
		identityUnit("W"),
	}},
	{ID: "potential", Units: []siUnitOption{identityUnit("V")}},
	{ID: "current", Units: []siUnitOption{identityUnit("A")}},
	{ID: "frequency", Units: []siUnitOption{
		// SignalK publishes revolutions in Hz; every tachometer shows RPM.
		linear("rpm", 60),
		identityUnit("Hz"),
	}},
	{ID: "raw", Units: []siUnitOption{identityUnit("raw")}},
	// Difference quantities, for a residual between two readings rather than
	// an absolute one (the twin-engine differential detector). A difference
	// in kelvin is numerically the same difference in Celsius, so unlike
	// "temperature" above, deltaK only scales -- it must never carry K's
	// -273.15 offset, or a 4 K gap would render as -269.1 degC. deltaPa
	// mirrors it in kPa, the unit an operator reads an oil or boost pressure
	// gap in, not "pressure"'s mb (which is for an absolute barometric
	// reading).
	{ID: "temperatureDelta", Units: []siUnitOption{identityUnit("deltaC"), identityUnit("deltaK")}},
	{ID: "pressureDelta", Units: []siUnitOption{linear("deltaKPa", 1.0/1000), identityUnit("deltaPa")}},
}

// convertToSI turns a value in a gauge's display unit into the SI value
// SignalK publishes. An unknown quantity or unit is an error rather than a
// pass-through: silently comparing psi against pascals is the failure this
// exists to prevent.
func convertToSI(value float64, quantityID, unitID string) (float64, error) {
	for _, q := range siQuantities {
		if q.ID != quantityID {
			continue
		}
		for _, u := range q.Units {
			if u.ID == unitID {
				return u.ToSI(value), nil
			}
		}
		return 0, fmt.Errorf("unknown unit %q for quantity %q", unitID, quantityID)
	}
	return 0, fmt.Errorf("unknown quantity %q", quantityID)
}
